File: test/python_examples/nwdb_service.py
from __future__ import annotations
import asyncio
from typing import Optional, List, Dict
import aiohttp
from bs4 import BeautifulSoup
from utils.cache import TTLCache
from .artifact_objective_service import build_artifact_objectives
import requests
import re
from urllib.parse import quote
import re as _re
NW_BUDDY_OBJECTIVE_TASKS_URL = "https://www.nw-buddy.de/nw-data/datatables/javelindata_objectivetasks.json"
NW_BUDDY_GAMEMODES_URL = "https://www.nw-buddy.de/nw-data/datatables/javelindata_gamemodes.json"

KNOWN_GAMEMODE_NAMES = {
    "DungeonAmrine": "Amrine Excavation",
    "DungeonShatteredObelisk": "Starstone Barrows",
    "DungeonRestlessShores01": "The Depths",
    "DungeonEbonscale00": "Dynasty Shipyard",
    "DungeonEdengrove00": "Garden of Genesis",
    "DungeonReekwater00": "Lazarus Instrumentality",
    "DungeonCutlassKeys00": "Barnacles and Black Powder",
    "DungeonShatterMtn00": "Tempest's Heart",
    "DungeonBrimstoneSands00": "The Ennead",
}

def _fallback_label(identifier: str) -> Optional[str]:
    token = (identifier or '').strip()
    if not token:
        return None
    token = re.sub(r"^(Dungeon|Activity)", "", token)
    token = re.sub(r"\d+$", "", token)
    spaced = re.sub(r"(?<!^)(?=[A-Z])", " ", token).strip()
    return spaced or None

class NWDBService:
    BASE = "https://nwdb.info"

    def __init__(self, session: aiohttp.ClientSession, cache: TTLCache | None = None):
        self.session = session
        self.cache = cache or TTLCache(3600)
        self._lock = asyncio.Lock()

    def _build_url(self, kind: str, slug: str) -> str:
        kind = kind.strip().lower()
        slug = slug.strip()
        return f"{self.BASE}/?{kind}={slug}"

    def fetch_nwdb_all(self):
        url = "https://nwdb.info/db/search/[[all]]"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                          "AppleWebKit/537.36 (KHTML, like Gecko) "
                          "Chrome/140.0.0.0 Safari/537.36",
            "Accept": "*/*",
        }
        try:
            response = requests.get(url, headers=headers, timeout=10)
            if response.status_code == 304:
                return {"status": "not_modified"}
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {"error": str(e)}

    def _normalize_slug(self, text: str) -> str:
        """
        Fallback slugifier in case API doesn't provide one.
        """
        s = re.sub(r"\s+", "-", text.strip().lower())
        s = re.sub(r"[^a-z0-9\-]+", "", s)
        return s

    def fetch_nwdb_items(self) -> List[Dict[str, str]]:
        """
        Returns a list of dicts: [{"id": "...", "name": "...", "slug": "..."}]
        """
        data = self.fetch_nwdb_all()
        items = []
        for item in (data.get("data", []) or []):
            if item.get("type") != "item":
                continue
            name = item.get("name")
            if not name:
                continue
            item_id = item.get("id")  # <- keep the canonical id
            slug = item.get("slug") or item_id or self._normalize_slug(name)
            items.append({"id": str(item_id or ""), "name": name, "slug": str(slug)})
        return items


    # --- New: cached item search for autocomplete ---
    def _get_cached_items(self) -> List[Dict[str, str]]:
        key = "nwdb_items_all"
        cached = self.cache.get(key)
        if cached is not None:
            return cached
        items = self.fetch_nwdb_items()
        # Cache for 1 hour (TTLCache given 3600s in __init__)
        self.cache.set(key, items)
        return items

    async def search_items(self, query: str, limit: int = 25) -> List[Dict[str, str]]:
        """
        Prefix-first, then substring matches, case-insensitive.
        Returns [{"name","slug"}] up to `limit`.
        """
        q = (query or "").strip().lower()
        if not q:
            return []
        items = self._get_cached_items()

        starts, contains = [], []
        for it in items:
            nm = (it.get("name") or "").lower()
            if nm.startswith(q):
                starts.append(it)
            elif q in nm:
                contains.append(it)

        ranked = starts + contains
        return ranked[: min(limit, 25)]


    def get_item_details(self, item_id: str):
        url = f"https://nwdb.info/db/item/{item_id}.json"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                          "AppleWebKit/537.36 (KHTML, like Gecko) "
                          "Chrome/140.0.0.0 Safari/537.36",
            "Accept": "*/*",
        }
        try:
            response = requests.get(url, headers=headers, timeout=10)
            if response.status_code == 304:
                return {"status": "not_modified"}
            response.raise_for_status()
            return response.json()
        except requests.RequestException as e:
            return {"error": str(e)}

    async def get_item_details_async(self, item_id: str):
        url = f"https://nwdb.info/db/item/{item_id}.json"
        try:
            timeout = aiohttp.ClientTimeout(total=10)
            async with self.session.get(url, timeout=timeout) as resp:
                if resp.status == 304:
                    return {"status": "not_modified"}
                if resp.status >= 400:
                    return {"error": f"HTTP {resp.status}"}
                return await resp.json()
        except Exception as e:
            return {"error": str(e)}

    # ---- Perks by item type (for /item_lookup) ----
    def _clean_text(self, text: str | None) -> str:
        if not text:
            return ""
        # Replace occurrences like ${20} or ${20.5} with just 20 or 20.5
        cleaned = _re.sub(r"\$\{([^}]+)\}", r"\1", text)
        # Collapse multiple spaces
        cleaned = _re.sub(r"\s+", " ", cleaned).strip()
        return cleaned
    async def _fetch_perks_page(self, item_type: str, page: int) -> dict:
        # NWDB expects lowercase values for filter_perk_item
        safe_item = quote((item_type or "").lower(), safe="")
        # Use the generic perks listing (not /attribute) and JSON endpoint
        url = f"https://nwdb.info/db/perks/page/{page}.json?filter_perk_item={safe_item}"
        print(f"[NWDB] GET {url}")
        timeout = aiohttp.ClientTimeout(total=10)
        async with self.session.get(url, timeout=timeout) as resp:
            if resp.status >= 400:
                return {"error": f"HTTP {resp.status}"}
            return await resp.json()

    async def fetch_all_perks_for_item_type(self, item_type: str) -> List[Dict]:
        item_type = (item_type or "").lower()
        key = f"perks:{item_type}"
        cached = self.cache.get(key)
        if cached is not None:
            return cached

        first = await self._fetch_perks_page(item_type, 1)
        if not first or first.get("error"):
            return []
        page_count = int(first.get("pageCount") or 1)
        perks = list(first.get("data") or [])
        if page_count > 1:
            # Fetch remaining pages concurrently
            tasks = [self._fetch_perks_page(item_type, p) for p in range(2, page_count + 1)]
            try:
                results = await asyncio.gather(*tasks, return_exceptions=True)
                for r in results:
                    if isinstance(r, dict):
                        perks.extend(r.get("data") or [])
            except Exception:
                pass

        # Cache condensed list for 30 minutes
        self.cache.set(key, perks, ttl=1800)
        return perks

    async def search_perks_for_item_type(self, item_type: str, query: str, limit: int = 25) -> List[Dict[str, str]]:
        q = (query or "").strip().lower()
        perks = await self.fetch_all_perks_for_item_type((item_type or "").lower())
        if not perks:
            return []
        starts, contains = [], []
        for p in perks:
            name = (p.get("name") or "").lower()
            if not name:
                continue
            if q and name.startswith(q):
                starts.append(p)
            elif q in name:
                contains.append(p)
        ranked = (starts + contains) if q else perks
        out = []
        for p in ranked[: min(limit, 25)]:
            out.append({
                "id": str(p.get("id") or ""),
                "name": p.get("name") or "",
                "description": self._clean_text(p.get("description")),
                "ScalingPerGearScore": p.get("ScalingPerGearScore"),
            })
        return out

    # ---- Items by perk id (for /item_lookup) ----
    async def fetch_items_by_perk_first_page(self, perk_id: str) -> List[Dict[str, str]]:
        # Only fetch page 1 as requested, then take first 5 entries
        safe_perk = quote(str(perk_id), safe="")
        url = f"https://nwdb.info/db/items/page/1.json?filter_perks={safe_perk}"
        print(f"[NWDB] GET {url}")
        timeout = aiohttp.ClientTimeout(total=10)
        async with self.session.get(url, timeout=timeout) as resp:
            if resp.status >= 400:
                return []
            payload = await resp.json()
        items = []
        for it in (payload.get("data") or [])[:5]:
            items.append({
                "id": str(it.get("id") or ""),
                "name": it.get("name") or "",
                "description": self._clean_text(it.get("description")),
            })
        return items

    async def fetch_items_by_perk_summary(self, perk_id: str) -> dict:
        """
        Fetch page 1 to get first 5 items and pageCount; if multiple pages, fetch last page
        to compute an exact total count without pulling every page.
        Returns: {"items": [...first5...], "pageCount": int, "total": int}
        """
        safe_perk = quote(str(perk_id), safe="")
        url1 = f"https://nwdb.info/db/items/page/1.json?filter_perks={safe_perk}"
        print(f"[NWDB] GET {url1}")
        timeout = aiohttp.ClientTimeout(total=10)
        async with self.session.get(url1, timeout=timeout) as resp:
            if resp.status >= 400:
                return {"items": [], "pageCount": 0, "total": 0}
            p1 = await resp.json()

        data1 = p1.get("data") or []
        page_count = int(p1.get("pageCount") or 1)
        per_page = len(data1)
        items = []
        for it in data1[:5]:
            items.append({
                "id": str(it.get("id") or ""),
                "name": it.get("name") or "",
                "description": self._clean_text(it.get("description")),
            })

        if page_count <= 1:
            total = per_page
            return {"items": items, "pageCount": page_count, "total": total}

        # Fetch last page just to count its entries
        urll = f"https://nwdb.info/db/items/page/{page_count}.json?filter_perks={safe_perk}"
        print(f"[NWDB] GET {urll}")
        async with self.session.get(urll, timeout=timeout) as resp2:
            if resp2.status >= 400:
                # Fall back to minimum estimate
                total = per_page * (page_count - 1)
                return {"items": items, "pageCount": page_count, "total": total}
            plast = await resp2.json()
        last_count = len(plast.get("data") or [])
        total = per_page * (page_count - 1) + last_count
        return {"items": items, "pageCount": page_count, "total": total}



    async def _fetch_objective_tasks(self) -> List[Dict]:
        key = "nwbuddy:objective_tasks"
        cached = self.cache.get(key)
        if cached is not None:
            return cached
        timeout = aiohttp.ClientTimeout(total=20)
        try:
            async with self.session.get(NW_BUDDY_OBJECTIVE_TASKS_URL, timeout=timeout) as resp:
                if resp.status >= 400:
                    self.cache.set(key, [], ttl=900)
                    return []
                payload = await resp.json()
        except Exception:
            return []
        tasks = payload.get("data") if isinstance(payload, dict) else payload
        if not isinstance(tasks, list):
            tasks = []
        self.cache.set(key, tasks, ttl=3600)
        return tasks

    async def _fetch_creature_lookup(self, creature_id: str) -> Optional[tuple[str, Optional[str]]]:
        if not creature_id:
            return None
        key = f"creature:{str(creature_id).lower()}"
        cached = self.cache.get(key)
        if cached is False:
            return None
        if cached:
            return cached
        url = f"{self.BASE}/db/creature/{creature_id}.json"
        try:
            timeout = aiohttp.ClientTimeout(total=10)
            async with self.session.get(url, timeout=timeout) as resp:
                if resp.status >= 400:
                    self.cache.set(key, False, ttl=900)
                    return None
                payload = await resp.json()
        except Exception:
            return None
        data = payload.get("data") if isinstance(payload, dict) else None
        name = None
        if isinstance(data, dict):
            name = data.get("Name") or data.get("name")
        if name:
            result = (str(name), f"{self.BASE}/db/creature/{creature_id}")
            self.cache.set(key, result, ttl=86400)
            return result
        self.cache.set(key, False, ttl=900)
        return None

    async def _fetch_zone_lookup(self, zone_id: str) -> Optional[tuple[str, Optional[str]]]:
        if not zone_id:
            return None
        zone_key = f"zone:{zone_id}"
        cached = self.cache.get(zone_key)
        if cached is False:
            return None
        if cached:
            return cached
        url = f"{self.BASE}/db/zone/{zone_id}.json"
        try:
            timeout = aiohttp.ClientTimeout(total=10)
            async with self.session.get(url, timeout=timeout) as resp:
                if resp.status >= 400:
                    self.cache.set(zone_key, False, ttl=900)
                    return None
                payload = await resp.json()
        except Exception:
            return None
        data = payload.get("data") if isinstance(payload, dict) else None
        name = None
        if isinstance(data, dict):
            name = data.get("name") or data.get("Name")
        if name:
            result = (str(name), f"{self.BASE}/db/zone/{zone_id}")
            self.cache.set(zone_key, result, ttl=86400)
            return result
        self.cache.set(zone_key, False, ttl=900)
        return None

    async def _load_gamemodes_index(self) -> List[Dict]:
        key = "nwbuddy:gamemodes"
        cached = self.cache.get(key)
        if cached is not None:
            return cached
        timeout = aiohttp.ClientTimeout(total=10)
        try:
            async with self.session.get(NW_BUDDY_GAMEMODES_URL, timeout=timeout) as resp:
                if resp.status >= 400:
                    self.cache.set(key, [], ttl=1800)
                    return []
                payload = await resp.json()
        except Exception:
            return []
        if not isinstance(payload, list):
            payload = []
        self.cache.set(key, payload, ttl=86400)
        return payload

    async def _fetch_gamemode_lookup(self, gamemode_id: str) -> Optional[tuple[str, Optional[str]]]:
        if not gamemode_id:
            return None
        gm_key = f"gamemode:{str(gamemode_id).lower()}"
        cached = self.cache.get(gm_key)
        if cached is False:
            return None
        if cached:
            return cached
        name = KNOWN_GAMEMODE_NAMES.get(gamemode_id)
        if not name:
            gamemodes = await self._load_gamemodes_index()
            entry = next((g for g in gamemodes if str(g.get("GameModeId")) == gamemode_id), None)
            if entry:
                display_key = entry.get("DisplayName")
                overrides = {
                    "@dungeon_amrine_title": "Amrine Excavation",
                    "@dungeon_everfall_title": "Starstone Barrows",
                    "@dungeon_restlessshores01_title": "The Depths",
                    "@dungeon_ebonscale00_title": "Dynasty Shipyard",
                    "@dungeon_edengrove00_title": "Garden of Genesis",
                    "@dungeon_reekwater_title": "Lazarus Instrumentality",
                    "@dungeon_cutlasskeys00_title": "Barnacles and Black Powder",
                    "@dungeon_shattermtn00_title": "Tempest''s Heart",
                    "@dungeon_brimstonesands00_title": "The Ennead",
                }
                if isinstance(display_key, str):
                    name = overrides.get(display_key)
                if not name:
                    script_name = entry.get("ScriptName")
                    if isinstance(script_name, str):
                        name = _fallback_label(script_name)
        if not name:
            name = _fallback_label(str(gamemode_id))
        result = (name, None) if name else None
        self.cache.set(gm_key, result if result else False, ttl=86400)
        return result

    async def fetch_artifact_objectives(self, item_id: str) -> List[str]:
        if not item_id:
            return []
        try:
            tasks = await self._fetch_objective_tasks()
        except Exception:
            return []
        if not tasks:
            return []

        async def _creature_lookup(creature: str) -> Optional[tuple[str, Optional[str]]]:
            return await self._fetch_creature_lookup(creature)

        async def _zone_lookup(zone: str) -> Optional[tuple[str, Optional[str]]]:
            return await self._fetch_zone_lookup(zone)

        async def _gamemode_lookup(gamemode: str) -> Optional[tuple[str, Optional[str]]]:
            return await self._fetch_gamemode_lookup(gamemode)

        try:
            return await build_artifact_objectives(item_id, tasks, _creature_lookup, _zone_lookup, _gamemode_lookup)
        except Exception:
            return []

File: services/nwdbService.ts
// NWDB Service - TypeScript implementation based on nwdb_service.py
// Handles NWDB API calls, caching, and data fetching

export interface NameLink {
  name: string;
  link: string | null;
}

export interface SearchResult {
  id: string;
  name: string;
  slug: string;
}

export interface PerkSearchResult {
  id: string;
  name: string;
  description: string;
  ScalingPerGearScore?: string;
}

export interface ItemSearchResult {
  id: string;
  name: string;
  description: string;
}

export interface PerkSummary {
  items: ItemSearchResult[];
  pageCount: number;
  total: number;
}

const NW_BUDDY_OBJECTIVE_TASKS_URL = "https://www.nw-buddy.de/nw-data/datatables/javelindata_objectivetasks.json";
const NW_BUDDY_GAMEMODES_URL = "https://www.nw-buddy.de/nw-data/datatables/javelindata_gamemodes.json";

const KNOWN_GAMEMODE_NAMES: Record<string, string> = {
  "DungeonAmrine": "Amrine Excavation",
  "DungeonShatteredObelisk": "Starstone Barrows",
  "DungeonRestlessShores01": "The Depths",
  "DungeonEbonscale00": "Dynasty Shipyard",
  "DungeonEdengrove00": "Garden of Genesis",
  "DungeonReekwater00": "Lazarus Instrumentality",
  "DungeonCutlassKeys00": "Barnacles and Black Powder",
  "DungeonShatterMtn00": "Tempest's Heart",
  "DungeonBrimstoneSands00": "The Ennead",
};

// Simple in-memory cache
class SimpleCache {
  private cache = new Map<string, { data: any; expires: number }>();
  private ttl: number;

  constructor(ttlSeconds: number = 3600) {
    this.ttl = ttlSeconds * 1000;
  }

  get(key: string): any {
    const item = this.cache.get(key);
    if (!item) return null;
    
    if (Date.now() > item.expires) {
      this.cache.delete(key);
      return null;
    }
    
    return item.data;
  }

  set(key: string, data: any, customTtl?: number): void {
    const ttl = customTtl ? customTtl * 1000 : this.ttl;
    this.cache.set(key, {
      data,
      expires: Date.now() + ttl
    });
  }

  clear(): void {
    this.cache.clear();
  }
}

export class NWDBService {
  private base = "https://nwdb.info";
  private cache = new SimpleCache(3600); // 1 hour default TTL

  private async fetchWithTimeout(url: string, timeoutMs: number = 10000): Promise<Response> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
    
    try {
      const response = await fetch(url, {
        signal: controller.signal,
        headers: {
          "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36",
          "Accept": "*/*",
        }
      });
      clearTimeout(timeoutId);
      return response;
    } catch (error) {
      clearTimeout(timeoutId);
      throw error;
    }
  }

  private buildUrl(kind: string, slug: string): string {
    const cleanKind = kind.trim().toLowerCase();
    const cleanSlug = slug.trim();
    return `${this.base}/?${cleanKind}=${cleanSlug}`;
  }

  public async fetchNWDBAll(): Promise<any> {
    const url = "https://nwdb.info/db/search/[[all]]";
    
    try {
      const response = await this.fetchWithTimeout(url);
      if (response.status === 304) {
        return { status: "not_modified" };
      }
      if (response.status >= 400) {
        return { error: `HTTP ${response.status}: ${response.statusText}` };
      }
      return await response.json();
    } catch (error) {
      return { error: String(error) };
    }
  }

  private normalizeSlug(text: string): string {
    return text
      .trim()
      .toLowerCase()
      .replace(/\s+/g, '-')
      .replace(/[^a-z0-9\-]+/g, '');
  }

  public async fetchNWDBItems(): Promise<SearchResult[]> {
    const data = await this.fetchNWDBAll();
    const items: SearchResult[] = [];
    
    const allItems = data.data || data;
    for (const item of allItems) {
      if (item.type !== "item" || !item.name) continue;
      
      const itemId = item.id;
      const slug = item.slug || itemId || this.normalizeSlug(item.name);
      
      items.push({
        id: String(itemId || ''),
        name: item.name,
        slug: String(slug)
      });
    }
    
    return items;
  }

  private getCachedItems(): SearchResult[] | null {
    return this.cache.get("nwdb_items_all");
  }

  private setCachedItems(items: SearchResult[]): void {
    this.cache.set("nwdb_items_all", items);
  }

  public async searchItems(query: string, limit: number = 25): Promise<SearchResult[]> {
    const q = (query || "").trim().toLowerCase();
    if (!q) return [];

    let items = this.getCachedItems();
    if (!items) {
      items = await this.fetchNWDBItems();
      this.setCachedItems(items);
    }

    const starts: SearchResult[] = [];
    const contains: SearchResult[] = [];

    for (const it of items) {
      const name = (it.name || "").toLowerCase();
      if (name.startsWith(q)) {
        starts.push(it);
      } else if (name.includes(q)) {
        contains.push(it);
      }
    }

    const ranked = [...starts, ...contains];
    return ranked.slice(0, Math.min(limit, 25));
  }

  public async getItemDetails(itemId: string): Promise<any> {
    const url = `https://nwdb.info/db/item/${itemId}.json`;
    
    try {
      const response = await this.fetchWithTimeout(url);
      if (response.status === 304) {
        return { status: "not_modified" };
      }
      if (response.status >= 400) {
        return { error: `HTTP ${response.status}: ${response.statusText}` };
      }
      return await response.json();
    } catch (error) {
      return { error: String(error) };
    }
  }

  private cleanText(text?: string): string {
    if (!text) return "";
    
    // Replace ${...} with just the content
    let cleaned = text.replace(/\$\{([^}]+)\}/g, "$1");
    // Collapse multiple spaces
    cleaned = cleaned.replace(/\s+/g, " ").trim();
    return cleaned;
  }

  private async fetchPerksPage(itemType: string, page: number): Promise<any> {
    const safeItem = encodeURIComponent((itemType || "").toLowerCase());
    const url = `https://nwdb.info/db/perks/page/${page}.json?filter_perk_item=${safeItem}`;
    
    try {
      const response = await this.fetchWithTimeout(url);
      if (response.status >= 400) {
        return { error: `HTTP ${response.status}: ${response.statusText}` };
      }
      return await response.json();
    } catch (error) {
      return { error: String(error) };
    }
  }

  public async fetchAllPerksForItemType(itemType: string): Promise<any[]> {
    const cleanItemType = (itemType || "").toLowerCase();
    const key = `perks:${cleanItemType}`;
    
    const cached = this.cache.get(key);
    if (cached !== null) return cached;

    const first = await this.fetchPerksPage(cleanItemType, 1);
    if (!first || first.error) return [];

    const pageCount = parseInt(first.pageCount || "1");
    let perks = [...(first.data || [])];

    if (pageCount > 1) {
      // Fetch remaining pages concurrently
      const tasks = Array.from({ length: pageCount - 1 }, (_, i) => 
        this.fetchPerksPage(cleanItemType, i + 2)
      );
      
      try {
        const results = await Promise.allSettled(tasks);
        for (const result of results) {
          if (result.status === 'fulfilled' && result.value && !result.value.error) {
            perks.push(...(result.value.data || []));
          }
        }
      } catch (error) {
        // Ignore errors and continue with what we have
      }
    }

    // Cache for 30 minutes
    this.cache.set(key, perks, 1800);
    return perks;
  }

  public async searchPerksForItemType(itemType: string, query: string, limit: number = 25): Promise<PerkSearchResult[]> {
    const q = (query || "").trim().toLowerCase();
    const perks = await this.fetchAllPerksForItemType((itemType || "").toLowerCase());
    
    if (!perks.length) return [];

    const starts: any[] = [];
    const contains: any[] = [];

    for (const p of perks) {
      const name = (p.name || "").toLowerCase();
      if (!name) continue;
      
      if (q && name.startsWith(q)) {
        starts.push(p);
      } else if (q && name.includes(q)) {
        contains.push(p);
      }
    }

    const ranked = q ? [...starts, ...contains] : perks;
    const results: PerkSearchResult[] = [];

    for (const p of ranked.slice(0, Math.min(limit, 25))) {
      results.push({
        id: String(p.id || ""),
        name: p.name || "",
        description: this.cleanText(p.description),
        ScalingPerGearScore: p.ScalingPerGearScore
      });
    }

    return results;
  }

  public async fetchItemsByPerkFirstPage(perkId: string): Promise<ItemSearchResult[]> {
    const safePerk = encodeURIComponent(String(perkId));
    const url = `https://nwdb.info/db/items/page/1.json?filter_perks=${safePerk}`;
    
    try {
      const response = await this.fetchWithTimeout(url);
      if (response.status >= 400) return [];
      
      const payload = await response.json();
      const items: ItemSearchResult[] = [];
      
      for (const it of (payload.data || []).slice(0, 5)) {
        items.push({
          id: String(it.id || ""),
          name: it.name || "",
          description: this.cleanText(it.description)
        });
      }
      
      return items;
    } catch (error) {
      return [];
    }
  }

  public async fetchItemsByPerkSummary(perkId: string): Promise<PerkSummary> {
    const safePerk = encodeURIComponent(String(perkId));
    const url1 = `https://nwdb.info/db/items/page/1.json?filter_perks=${safePerk}`;
    
    try {
      const response1 = await this.fetchWithTimeout(url1);
      if (response1.status >= 400) {
        return { items: [], pageCount: 0, total: 0 };
      }
      
      const p1 = await response1.json();
      const data1 = p1.data || [];
      const pageCount = parseInt(p1.pageCount || "1");
      const perPage = data1.length;
      
      const items: ItemSearchResult[] = [];
      for (const it of data1.slice(0, 5)) {
        items.push({
          id: String(it.id || ""),
          name: it.name || "",
          description: this.cleanText(it.description)
        });
      }

      if (pageCount <= 1) {
        return { items, pageCount, total: perPage };
      }

      // Fetch last page to get exact count
      const urlLast = `https://nwdb.info/db/items/page/${pageCount}.json?filter_perks=${safePerk}`;
      try {
        const responseLast = await this.fetchWithTimeout(urlLast);
        if (responseLast.status >= 400) {
          // Fall back to minimum estimate
          return { items, pageCount, total: perPage * (pageCount - 1) };
        }
        
        const plast = await responseLast.json();
        const lastCount = (plast.data || []).length;
        const total = perPage * (pageCount - 1) + lastCount;
        
        return { items, pageCount, total };
      } catch (error) {
        // Fall back to minimum estimate
        return { items, pageCount, total: perPage * (pageCount - 1) };
      }
    } catch (error) {
      return { items: [], pageCount: 0, total: 0 };
    }
  }

  private async fetchObjectiveTasks(): Promise<any[]> {
    const key = "nwbuddy:objective_tasks";
    const cached = this.cache.get(key);
    if (cached !== null) return cached;

    try {
      const response = await this.fetchWithTimeout(NW_BUDDY_OBJECTIVE_TASKS_URL, 20000);
      if (response.status >= 400) {
        this.cache.set(key, [], 900);
        return [];
      }
      
      const payload = await response.json();
      const tasks = payload.data || payload;
      
      if (!Array.isArray(tasks)) {
        this.cache.set(key, [], 3600);
        return [];
      }
      
      this.cache.set(key, tasks, 3600);
      return tasks;
    } catch (error) {
      return [];
    }
  }

  private async fetchCreatureLookup(creatureId: string): Promise<NameLink | null> {
    if (!creatureId) return null;
    
    const key = `creature:${String(creatureId).toLowerCase()}`;
    const cached = this.cache.get(key);
    if (cached === false) return null;
    if (cached) return cached;

    const url = `${this.base}/db/creature/${creatureId}.json`;
    
    try {
      const response = await this.fetchWithTimeout(url);
      if (response.status >= 400) {
        this.cache.set(key, false, 900);
        return null;
      }
      
      const payload = await response.json();
      const data = payload.data || payload;
      const name = data?.Name || data?.name;
      
      if (name) {
        const result: NameLink = {
          name: String(name),
          link: `${this.base}/db/creature/${creatureId}`
        };
        this.cache.set(key, result, 86400);
        return result;
      }
      
      this.cache.set(key, false, 900);
      return null;
    } catch (error) {
      return null;
    }
  }

  private async fetchZoneLookup(zoneId: string): Promise<NameLink | null> {
    if (!zoneId) return null;
    
    const key = `zone:${zoneId}`;
    const cached = this.cache.get(key);
    if (cached === false) return null;
    if (cached) return cached;

    const url = `${this.base}/db/zone/${zoneId}.json`;
    
    try {
      const response = await this.fetchWithTimeout(url);
      if (response.status >= 400) {
        this.cache.set(key, false, 900);
        return null;
      }
      
      const payload = await response.json();
      const data = payload.data || payload;
      const name = data?.name || data?.Name;
      
      if (name) {
        const result: NameLink = {
          name: String(name),
          link: `${this.base}/db/zone/${zoneId}`
        };
        this.cache.set(key, result, 86400);
        return result;
      }
      
      this.cache.set(key, false, 900);
      return null;
    } catch (error) {
      return null;
    }
  }

  private async loadGamemodesIndex(): Promise<any[]> {
    const key = "nwbuddy:gamemodes";
    const cached = this.cache.get(key);
    if (cached !== null) return cached;

    try {
      const response = await this.fetchWithTimeout(NW_BUDDY_GAMEMODES_URL);
      if (response.status >= 400) {
        this.cache.set(key, [], 1800);
        return [];
      }
      
      const payload = await response.json();
      const data = Array.isArray(payload) ? payload : [];
      this.cache.set(key, data, 86400);
      return data;
    } catch (error) {
      return [];
    }
  }

  private fallbackLabel(identifier: string): string | null {
    const token = (identifier || '').trim();
    if (!token) return null;
    
    let cleaned = token
      .replace(/^(Dungeon|Activity)/, '')
      .replace(/\d+$/, '');
    
    cleaned = cleaned.replace(/(?<!^)(?=[A-Z])/g, ' ').trim();
    return cleaned || null;
  }

  private async fetchGamemodeLookup(gamemodeId: string): Promise<NameLink | null> {
    if (!gamemodeId) return null;
    
    const key = `gamemode:${String(gamemodeId).toLowerCase()}`;
    const cached = this.cache.get(key);
    if (cached === false) return null;
    if (cached) return cached;

    let name = KNOWN_GAMEMODE_NAMES[gamemodeId];
    
    if (!name) {
      const gamemodes = await this.loadGamemodesIndex();
      const entry = gamemodes.find(g => String(g.GameModeId) === gamemodeId);
      
      if (entry) {
        const displayKey = entry.DisplayName;
        const overrides: Record<string, string> = {
          "@dungeon_amrine_title": "Amrine Excavation",
          "@dungeon_everfall_title": "Starstone Barrows",
          "@dungeon_restlessshores01_title": "The Depths",
          "@dungeon_ebonscale00_title": "Dynasty Shipyard",
          "@dungeon_edengrove00_title": "Garden of Genesis",
          "@dungeon_reekwater_title": "Lazarus Instrumentality",
          "@dungeon_cutlasskeys00_title": "Barnacles and Black Powder",
          "@dungeon_shattermtn00_title": "Tempest's Heart",
          "@dungeon_brimstonesands00_title": "The Ennead",
        };
        
        if (typeof displayKey === 'string') {
          name = overrides[displayKey];
        }
        
        if (!name) {
          const scriptName = entry.ScriptName;
          if (typeof scriptName === 'string') {
            name = this.fallbackLabel(scriptName);
          }
        }
      }
    }
    
    if (!name) {
      name = this.fallbackLabel(String(gamemodeId));
    }
    
    const result = name ? { name, link: null } : null;
    this.cache.set(key, result || false, 86400);
    return result;
  }

  public async fetchArtifactObjectives(itemId: string): Promise<string[]> {
    if (!itemId) return [];
    
    try {
      const tasks = await this.fetchObjectiveTasks();
      if (!tasks.length) return [];

      // Import artifact objective service
      const { buildArtifactObjectives } = await import('./artifactObjectiveService');
      
      return await buildArtifactObjectives(
        itemId,
        tasks,
        (creature: string) => this.fetchCreatureLookup(creature),
        (zone: string) => this.fetchZoneLookup(zone),
        (gamemode: string) => this.fetchGamemodeLookup(gamemode)
      );
    } catch (error) {
      return [];
    }
  }
}

export const nwdbService = new NWDBService();
